# Week 4 reading report
Read README.md. Keep units in your explanation.


<!-- assignment-workflow -->
## Before you begin
Open ASSIGNMENT_INSTRUCTIONS.txt from your extracted folder for the starting files, steps, checks and exact upload list. Read STUDENT_HOW_TO.md in a text editor for account creation and Colab/download help. Keep the required names below unchanged. Write your own answers; guided_practice.ipynb is not a graded upload.


In [1]:
STUDENT_ID="24/UG/654/BSCE-J"
COHORT="CML3112"
import csv
with open("site_inspection_log.csv",newline="") as f:
    rows=list(csv.DictReader(f))


## valid_count (5 marks)
Store an integer count of rows whose crack_width_mm field is nonblank. Skip blank readings before numeric conversion.


In [2]:
valid_count = 0
for row in rows:
  if row["crack_width_mm"] != "":
    valid_count += 1
print(valid_count)

297


Interesting Pattern
The final output is 297 instead of 300 due to the specific evaluation of the code's conditional statement: • The Evidence: The loop contains the condition if row["crack_width_mm"] != "":. This explicit logical check only allows the counter to increment (valid_count += 1) if the value inside the crack_width_mm column is not an empty string. • The Reason for 297: Out of the 300 total entries processed by the for loop, exactly 3 rows failed this specific condition because their crack_width_mm field matched "" (blank). As a result, the code bypassed the counter for those rows, yielding a final verified count of 297 non-blank entries.

## mean_reading (5 marks)
Store the arithmetic mean of the nonblank crack_width_mm values as a number. Add the valid readings and divide by their count. Keep full precision.


In [3]:
total_width = 0
for row in rows:
  if row["crack_width_mm"] != "":
    total_width += float(row["crack_width_mm"])
mean_reading = total_width / valid_count
print(mean_reading)


0.2262996632996631


Interesting Pattern
This code calculates the arithmetic mean (average) of all valid, non-blank crack width readings. • How it works: • Initialization: It sets total_width = 0 to accumulate the sum of the readings. • Filtering and Summing: The for loop checks each row. If the crack_width_mm field is not blank (!= ""), it converts the string value to a decimal number using float() and adds it to total_width. • Calculation: Outside the loop, it divides the accumulated total_width by the valid_count (the 297 valid rows calculated in the previous step). • The Result: The mean crack width across all valid samples is approximately 0.2263 mm

## above_limit (5 marks)
Store an integer count of valid crack_width_mm readings strictly greater than 0.3. A reading equal to the threshold is not counted.


In [4]:
above_limit = 0
for row in rows:
  if row["crack_width_mm"] != "" and float(row["crack_width_mm"]) > 0.3:
    above_limit += 1
print(above_limit)



59


Interesting Pattern
This code counts how many valid crack width measurements are strictly greater than a safety or design threshold of 0.3 mm. • How it works: • Initialization: It initializes a counter above_limit = 0. • Compound Condition: The if statement uses a logical and to check two things simultaneously: 1. row["crack_width_mm"] != "" — Ensures the data isn't blank to prevent a conversion error. 2. float(row["crack_width_mm"]) > 0.3 — Converts the valid string to a float and checks if it is strictly greater than 0.3. • Aggregation: If both conditions are met, the counter increments by 1. • The Result: There are exactly 59 entries in the dataset that exceed the 0.3 mm threshold.

## group_means (5 marks)
Store a dictionary whose keys are the structure_id labels and whose values are the mean valid crack_width_mm readings within each group. Divide each group total by its own valid count. Keep full precision.


In [5]:
group_totals = {}
group_counts = {}

# Accumulate totals and counts for each group, skipping blank values
for row in rows:
    structure_id = row['structure_id']
    crack_width_raw = row['crack_width_mm']
    if crack_width_raw != '':
        val = float(crack_width_raw)
        if structure_id not in group_totals:
            group_totals[structure_id] = 0.0
            group_counts[structure_id] = 0
        group_totals[structure_id] += val
        group_counts[structure_id] += 1

# Compute the mean for each group with full precision
group_means = {}
for struct_id in group_totals:
    group_means[struct_id] = group_totals[struct_id] / group_counts[struct_id]

print(group_means)


{'B1': 0.11067857142857143, 'B2': 0.1464642857142857, 'B3': 0.16533333333333336, 'B4': 0.1781666666666667, 'B5': 0.19691304347826086, 'B6': 0.2158333333333333, 'B7': 0.23891666666666664, 'B8': 0.26029166666666664, 'B9': 0.278375, 'B10': 0.303375, 'B11': 0.32030434782608697, 'B12': 0.3438333333333334}


Interesting Pattern
This code groups the data by structural identifiers (structure_id) and calculates a separate average crack width for each unique structure. • How it works: • Dynamic Dictionaries: Instead of flat variables, it initializes two empty dictionaries: group_totals (to store the sum of widths per group) and group_counts (to store the number of valid records per group). • Dynamic Group Initialization: Inside the loop, if a structure_id is encountered for the first time, it automatically creates entry keys for that group and sets their baseline values to 0.0 and 0. • Group-by Accumulation: It adds the float value to that specific structure's total and increments that specific structure's count. • Final Group Math: A final loop goes through each unique structure found, dividing its total width by its total count to compute a highly precise individual average. • The Result: It prints a dictionary mapping each structural label (like 'B5', 'B6', 'B7') directly to its specific average crack width.